packages/zoltan2/src/problems/Zoltan2_OrderingSolution.hpp

```diff
@@ -312,47 +308,93 @@ template <typename lno_t, typename gno_t>
     // than other arrays.
     return separatorRange_.getRawPtr();
   }
 
-  /*! \brief Get pointer to (local) separator tree.
+  /*! \brief Get pointer to separator tree.
    */
-  inline lno_t *getSeparatorTreeView() const
+  inline index_t *getSeparatorTreeView() const
   {
     if (perm_size_)
       return separatorTree_.getRawPtr();
     else
       return NULL;
   }
   
-  /*! \brief Get reference to (local) separator column block.
+  /*! \brief Get reference to separator column block.
    */
-  inline lno_t &NumSeparatorBlocks()
+  index_t &NumSeparatorBlocks()
   {
     return separatorColBlocks_; 
   }
 
+  /*! \brief returns 0 if permutation is valid, negative if invalid.
+   */
+  int validatePerm()
+  {
+    size_t n = getPermutationSize();
+
+    if(!havePerm_) {
+      return -1;
+    }
+
+    std::vector<int> count(getPermutationSize());
+    for (size_t i = 0; i < n; i++) {
+      count[i] = 0;
+    }
+
+    for (size_t i = 0; i < n; i++) {
+      if ( (perm_[i] < 0) || (perm_[i] >= static_cast<index_t>(n)) )
+        return -1;
+      else
+        count[perm_[i]]++;
+    }
+
+    // Each index should occur exactly once (count==1)
+    for (size_t i = 0; i < n; i++) {
+      if (count[i] != 1){
+        return -2;
+        break;
+      }
+    }
+
+    return 0;
+  }
+
 protected:
   // Ordering solution consists of permutation vector(s).
   // Either perm or invperm should be computed by the algorithm.
   size_t perm_size_;
-  // For now, assume permutations are local. Revisit later (e.g., for Scotch)
-  bool havePerm_;           // has perm_ been computed yet?
-  bool haveInverse_;        // has invperm_ been computed yet?
+  bool havePerm_;                    // has perm_ been computed yet?
+  bool haveInverse_;                 // has invperm_ been computed yet?
   bool haveSeparatorRange_;          // has sepRange_ been computed yet?
-  bool haveSeparatorTree_;        // has sepTree_ been computed yet?
-  ArrayRCP<lno_t> perm_;    // zero-based local permutation
-  ArrayRCP<lno_t> invperm_; // inverse of permutation above
-  ArrayRCP<lno_t> separatorRange_;   // range iterator for separator tree
-  ArrayRCP<lno_t> separatorTree_; // separator tree
-  lno_t separatorColBlocks_;        // number of column blocks in separator
+  bool haveSeparatorTree_;           // has sepTree_ been computed yet?
+  ArrayRCP<index_t> perm_;           // zero-based local permutation
+  ArrayRCP<index_t> invperm_;        // inverse of permutation above
+  ArrayRCP<index_t> separatorRange_; // range iterator for separator tree
+  ArrayRCP<index_t> separatorTree_;  // separator tree
+  index_t separatorColBlocks_;       // number of column blocks in separator
+};
+
+template <typename lno_t>
+class LocalOrderingSolution : public OrderingSolution<lno_t>
+{
+public:
+  LocalOrderingSolution(lno_t perm_size) : OrderingSolution<lno_t>(perm_size) {}
+};
+
+template <typename gno_t>
```

Comment:
We discussed both possibilities and decided to keep them separate to avoid having to determine how to act on local/global by reading the type. At this stage it would be easy to convert.